### ISSUE TITLE
iOS build errors in RCTLinkingManager.h when running react-native run-ios with Xcode 9.2

### ISSUE THREAD

**Issue by noxee** [facebook/react-native#20797]
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
```
React Native Environment Info:
  System:
    OS: macOS High Sierra 10.13.3
    CPU: x64 Intel(R) Core(TM) i7-3615QM CPU @ 2.30GHz
    Memory: 678.91 MB / 16.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 8.11.1 - ~/.nvm/versions/node/v8.11.1/bin/node
    npm: 4.6.1 - ~/.nvm/versions/node/v8.11.1/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 11.2, macOS 10.13, tvOS 11.2, watchOS 4.2
    Android SDK:
      Build Tools: 23.0.1, 27.0.3
      API Levels: 23, 27
  IDEs:
    Android Studio: 3.1 AI-173.4670197
    Xcode: 9.2/9C40b - /usr/bin/xcodebuild
  npmPackages:
    react: 16.4.1 => 16.4.1
    react-native: 0.57.0-rc.0 => 0.57.0-rc.0
  npmGlobalPackages:
    react-native-cli: 2.0.1
```

## Description
Using Xcode 9.2 if I run the command `react-native run-ios` I get the following build errors:
```
In file included from TestProject/node_modules/react-native/Libraries/LinkingIOS/RCTLinkingManager.m:8:
TestProject/node_modules/react-native/Libraries/LinkingIOS/RCTLinkingManager.h:14:36: error: pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Werror,-Wnullability-completeness]
+ (BOOL)application:(UIApplication *)app
                                   ^
TestProject/node_modules/react-native/Libraries/LinkingIOS/RCTLinkingManager.h:14:36: note: insert '_Nullable' if the pointer may be null
+ (BOOL)application:(UIApplication *)app
                                   ^
                                    _Nullable
TestProject/node_modules/react-native/Libraries/LinkingIOS/RCTLinkingManager.h:14:36: note: insert '_Nonnull' if the pointer should never be null
+ (BOOL)application:(UIApplication *)app
                                   ^
                                    _Nonnull
TestProject/node_modules/react-native/Libraries/LinkingIOS/RCTLinkingManager.h:25:29: error: block pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Werror,-Wnullability-completeness]
  restorationHandler:(void (^)(NSArray * __nullable))restorationHandler;
                            ^
TestProject/node_modules/react-native/Libraries/LinkingIOS/RCTLinkingManager.h:25:29: note: insert '_Nullable' if the block pointer may be null
  restorationHandler:(void (^)(NSArray * __nullable))restorationHandler;
                            ^
                             _Nullable
TestProject/node_modules/react-native/Libraries/LinkingIOS/RCTLinkingManager.h:25:29: note: insert '_Nonnull' if the block pointer should never be null
  restorationHandler:(void (^)(NSArray * __nullable))restorationHandler;
                            ^
                             _Nonnull
```
I believe it's related to the setting **Implicit Non-Literal Null Conversions** setting in Xcode 9.

## Reproducible Demo
```
react-native init TestProject --version="0.57.0-rc.0"
cd TestProject
react-native run-ios
```

I've tested updating the method signatures in the header file and it appears to build correctly, I'm going to create a pull request shortly with the changes but I wanted to create an issue incase anyone else has been experiencing this.

**Comment by keithics** [comment]
React Native Environment Info:
    System:
      OS: macOS Sierra 10.12.6
      CPU: x64 Intel(R) Core(TM) i5-2500 CPU @ 3.30GHz
      Memory: 389.90 MB / 16.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 9.11.2 - /usr/local/bin/node
      npm: 5.6.0 - /usr/local/bin/npm
      Watchman: 4.7.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 11.2, macOS 10.13, tvOS 11.2, watchOS 4.2
      Android SDK:
        Build Tools: 23.0.1, 23.0.2, 23.0.3, 24.0.3, 25.0.0, 25.0.1, 25.0.2, 25.0.3, 26.0.1, 26.0.2, 27.0.2, 27.0.3, 28.0.1
        API Levels: 22, 23, 24, 25, 26, 27, 28
    IDEs:
      Android Studio: 3.1 AI-173.4819257
      Xcode: 9.2/9C40b - /usr/bin/xcodebuild
    npmPackages:
      react: 16.0.0-alpha.12 => 16.0.0-alpha.12
      react-native: 0.57 => 0.57.0
    npmGlobalPackages:
      react-native-cli: 2.0.1

**Comment by markharding** [comment]
Can confirm this fix worked for me!

**Comment by diegolmello** [comment]
iOS will never build without this PR.
I know isn't merged yet, but it would be nice to have this fix in 0.57.1.
@hramos @kelset @grabbou Can we get attention to this, please?
Thank you.

**Comment by kelset** [comment]
Closing this as duplicate of #21071 since in that there are more info around the issue.
